### PR TITLE
chore(cli): apply allow-downgrade to env deploy/package

### PR DESCRIPTION
**Pull request by iamhopaul123**
Also fixes integ tests and init e2e test suite

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the Apache 2.0 License.

### CODE REVIEW

File: internal/pkg/cli/env_package.go [hunk 1]
```diff
@@ -148,16 +163,25 @@ func (o *packageEnvOpts) Ask() error {
 	if _, err := o.getAppCfg(); err != nil {
 		return err
 	}
-	return o.validateOrAskEnvName()
+	return o.validateOrAskname()
```
Comment (Lou1415926) [inline internal/pkg/cli/env_package.go]: nit:
```suggestion
	return o.validateOrAskName()
```

File: internal/pkg/deploy/cloudformation/cloudformation_integration_test.go [hunk 2]
```diff
@@ -55,6 +55,7 @@ func Test_App_Infrastructure(t *testing.T) {
 	deployer := cloudformation.New(sess, cloudformation.WithProgressTracker(os.Stderr))
 	cfClient := awsCF.New(sess)
 	require.NoError(t, err)
+	version.Version = "v1.28.0"
```
Comment (Lou1415926) [inline internal/pkg/deploy/cloudformation/cloudformation_integration_test.go:58]: what about we just nuke this file 🤔

File: internal/pkg/cli/env_deploy.go [hunk 3]
```diff
@@ -133,8 +149,32 @@ func (o *deployEnvOpts) Ask() error {
 	return o.validateOrAskEnvName()
 }
 
+func validateEnvVersion(vg versionGetter, name, templateVersion string) error {
+	envVersion, err := vg.Version()
```
Comment (Lou1415926) [inline internal/pkg/cli/env_deploy.go:153]: what about if the version is "bootstrap"?

File: internal/pkg/version/version.go [hunk 4]
```diff
@@ -11,6 +11,10 @@ const (
 	AppTemplateMinAlias = "v1.0.0"
 	// AppTemplateMinStaticSite is the minimum app version required to deploy a static site.
 	AppTemplateMinStaticSite = "v1.2.0"
+	// LegacyEnvTemplate is the version associated with the environment template before we started versioning.
+	LegacyEnvTemplate = "v0.0.0"
+	// EnvTemplateBootstrap is the version of an environment template that contains only bootstrap resources.
+	EnvTemplateBootstrap = "bootstrap"
```
Comment (Lou1415926) [inline internal/pkg/version/version.go:17]: A small call out which I think is probably fine - 
When people run `env init`, we deploy an "environment template" that contains only the CFN execution role and the env manager role. These two roles are used when people run `env deploy`.

An "environment template" at this point contains only the two roles and doesn't not contain any actual environment resources. This template is on "bootstrap" at this point.

With the current implementation, we (probably) won't be able to detect a "downgrade" in the following case:
```
$ copilot-v1.30.0 env init # Deploy an template on "bootstrap".
$ copilot-v1.29.0 env deploy 
```

Which I think is probably fine in most of the cases - unless we've patched some important permissions two either CFNExecutionRole or EnvManagerRole on v1.30.0.

File: internal/pkg/deploy/cloudformation/cloudformation_integration_test.go (hunk 2 above)
Comment (iamhopaul123) [reply internal/pkg/deploy/cloudformation/cloudformation_integration_test.go:58]: Yeah I am inclined too but I feel like we need at least a separate PR to remove the integration test.

File: internal/pkg/cli/env_deploy.go (hunk 3 above)
Comment (iamhopaul123) [reply internal/pkg/cli/env_deploy.go:153]: Good call we need to deal with this case too.

File: internal/pkg/version/version.go (hunk 4 above)
Comment (iamhopaul123) [reply internal/pkg/version/version.go:17]: Yeah I think it should be fine because `env init` won't function anyway.

**Review by huanjani — APPROVED** [review]
lgtm

**Review by Lou1415926 — APPROVED** [review]

**Pull request merged**